Fill in multiproof feature and  VDAF overview (#69)
junyechen1996/draft-chen-cfrg-vdaf-pine: draft-chen-cfrg-vdaf-pine.md

In order for the Client to shard its gradient into secret shares for the
Aggregators:
1. Each Client first encodes its gradient and the range-checked L2-norm results
   according to {{encode-gradient-and-norm}}.
1. Then it derives the wraparound joint randomness to run wraparound checks. It
   computes a wraparound joint randomness "part" for each Aggregator, by
   computing a XOF seed that hashes the secret share of the results from step 1
   for that Aggregator. Note the parts are sent to all Aggregators. The Client
   then computes a XOF seed that hashes all the parts and uses that to generate
   the random vectors in wraparound checks.
1. Then it encodes results for wraparound checks according to
   {{encode-wr-check}}, based on the encoded gradient from step 1, and the XOF
   seed from step 2. The results are appended at the end of the encoded gradient
   and range-checked norm from step 1, to form the fully encoded measurement
   that will be secret shared and sent to the Aggregators.
1. Then it needs to derive the verification joint randomness to run the validity
   circuit. The joint randomness part derivation and seed derivation are similar

>    to those for wraparound joint randomness, but the hash function for

Reviewer: what is the "hash function"?

   verification joint randomness needs to include the secret shares of the fully
   encoded measurement from step 3.
1. Finally, it generates the proof with the verification joint randomness from
   step 4, secret shares the proof, and sends the secret-shared encoded
   measurement and proof to the Aggregators.

Then the Aggregators carry out a multi-party computation to obtain the output
shares (the secret shares of the encoded Client gradients), and also reject
Client gradients that have invalid L2-norm:
1. The Aggregators first need to derive wraparound joint randomness in order to

>    run wraparound checks. Each Aggregator derives the XOF seed from its secret

Reviewer: ```suggestion
   run wraparound checks. Each Aggregator derives an XOF seed --- it's wraparound randomness "part" --- from its secret
```

Author: Reworded this part.

> TODO Specify the implementation of `Valid` from {{Section 7.3.2 of !VDAF}}.

> ## Multiple Proofs {#multiproof}

Author: Moved it to the VDAF section.

# The PINE VDAF {#vdaf}

This section describes PINE VDAF for {{ROCT23}}, a one-round VDAF with no
aggregation parameter. It takes a set of Client gradients expressed as vectors
of floating point values, and computes an element-wise summation of valid
gradients with bounded L2-norm configured by the user parameters in
{{pine-user-param}}. The VDAF largely uses the encoding and validation schemes
in {{flp}}, and also specifies how the joint randomness shared between the
Client and Aggregators is derived. There are two kinds of joint randomness used:

* "Verification joint randomness": These are the field elements used by the
  Client and Aggregators to evaluate the FLP circuit. The verification joint
  randomness is derived similar to the joint randomness in Prio3
  {{Section 7.2.1.2 of !VDAF}}: the XOF is applied to each secret share of the
  encoded measurement to derive the "part"; and the parts are hashed together,
  using the XOF once more, to get the seed for deriving the joint randomness
  itself.
* "Wraparound joint randomness": This is used to generate the random vectors in
  the wraparound checks that both the Clients and Aggregators need to derive on

>   their own. It is generated in much the same way as verification joint

Reviewer: ```suggestion
  their own. It is generated in much the same way as the verification joint
```

  randomness, except that only the gradient and the range-checked norm are used
  to derive the parts.

In order for the Client to shard its gradient into input shares for the
Aggregators, the Client first encodes its gradient into field elements, and
encodes the range-checked L2-norm, according to {{encode-gradient-and-norm}}.

> Then it needs to derives the wraparound joint randomness for the wraparound

Reviewer: ```suggestion
Next, it derives the wraparound joint randomness for the wraparound
```

> checks. It first derives the parts, by applying the XOF to the secret shares of
> the encoded gradient and range-checked norm. Then it applies the XOF again to
> the parts to obtain the seed for the wraparound joint randomness and use it to
> encode the results for wraparound checks {{encode-wr-check}}. The encoded

Reviewer: You've already said this above.
```suggestion
checks as described above. The encoded
```

gradient, range-checked norm, and results for wraparound checks form the fully

> encoded measurement, that will be secret-shared to (1) be sent as input shares

Reviewer: ```suggestion
input to the FLP circuit, that will be secret-shared to (1) be sent as input shares
```

> encoded measurement, that will be secret-shared to (1) be sent as input shares

Reviewer: ```suggestion
input to the FLP circuit, that will be secret-shared to (1) be sent as input shares
```

> for the Aggregators, and (2) derive the verification joint randomness parts.
> The parts are then hashed by the XOF to generate the seed for the verification
> joint randomness to evaluate the FLP circuit and generate the proof. The proof

Reviewer: ```suggestion
for the Aggregators, and (2) derive the verification joint randomness as described above. The proof
```

is also secret-shared, and sent to the Aggregators, along with the input shares,
and the joint randomness parts for both wraparound and verification joint
randomness.

Then the Aggregators carry out a multi-party computation to obtain the output

> shares (the secret shares of the encoded Client gradients), and also reject

Reviewer: ```suggestion
shares (the secret shares of the encoded Client gradient), and also reject
```

Client gradients that have invalid L2-norm. Each Aggregator first needs to

> derive wraparound and verification joint randomness. To compute the XOF seed

Reviewer: Like the last paragraph, avoid saying things twice. Make sure to highlight the differences between what the Client computes versus what each Aggregator computes.

Author: Done.

for wraparound joint randomness part, it applies its share of encoded gradient
and the range-checked norm to the XOF. Then it applies its part and other
Aggregators' parts to the XOF again to obtain the seed for wraparound joint

> for other Aggregators  , and also the wraparound joint randomness parts for the

Reviewer: ```suggestion
for other Aggregators, and also the wraparound joint randomness parts for the
```

other Aggregators. The verification joint randomness is derived similarly,
except each Aggregator uses its entire input share to derive the part. Then
each Aggregator runs the wraparound checks with the wraparound joint randomness,
and queries the FLP with its input share, proof share, the wraparound check
results, and the verification joint randomness. All Aggregators then exchange
the results from the FLP and decide whether to accept that Client gradient.

> Next, each Aggregator sums up the valid output shares locally, which are
> expressed as arrays of field elements. Finally, the Collector sums up the

Reviewer: ```suggestion
Next, each Aggregator sums up their shares of the encoded gradients and sends the aggregate share to the Collector. Finally, the Collector sums up the
```

aggregate shares to obtain the aggregate result, and decodes it into an array
of floating point values.

Like Prio3 {{Section 7.1.2 of !VDAF}}, PINE supports generation and verification

> of multiple FLPs. The goal is to reduce soundness error (Corollary 3.13 in

Reviewer: ```suggestion
of multiple FLPs. The goal is to improve robustness of PINE (Corollary 3.13 in
```
